contracts/src/v0.8/dev/transmission/4337/Paymaster.sol

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import "../../vendor/entrypoint/interfaces/IPaymaster.sol";
+import "./SCALibrary.sol";
+import "../../vendor/entrypoint/core/Helpers.sol";
+import "../../../interfaces/LinkTokenInterface.sol";
+import "../../../interfaces/AggregatorV3Interface.sol";
+import "./SCALibrary.sol";
+
+/// @dev LINK token paymaster implementation.
+/// TODO: more documentation.
+contract Paymaster is IPaymaster {
+  error OnlyCallableFromLink();
+  error InvalidCalldata();
+  error UserOperationAlreadyTried(bytes32 userOpHash);
+  error InsufficientFunds(uint256 juelsNeeded, uint256 subscriptionBalance);
+
+  LinkTokenInterface public immutable LINK;
+  AggregatorV3Interface public immutable LINK_ETH_FEED;
+
+  struct Config {
+    uint32 stalenessSeconds;
+    int256 fallbackWeiPerUnitLink;
+  }
+  Config public s_config;
+
+  mapping(bytes32 => bool) userOpHashMapping;
+  mapping(address => uint256) subscriptions;
+
+  constructor(LinkTokenInterface linkToken, AggregatorV3Interface linkEthFeed) {
+    LINK = linkToken;
+    LINK_ETH_FEED = linkEthFeed;
+  }
+
+  function setConfig(uint32 stalenessSeconds, int256 fallbackWeiPerUnitLink) external {
+    s_config = Config({stalenessSeconds: stalenessSeconds, fallbackWeiPerUnitLink: fallbackWeiPerUnitLink});
+  }
+
+  function onTokenTransfer(address /* _sender */, uint256 _amount, bytes calldata _data) external {
+    if (msg.sender != address(LINK)) {
+      revert OnlyCallableFromLink();
+    }
+    if (_data.length != 32) {
+      revert InvalidCalldata();
+    }
+
+    address subscription = abi.decode(_data, (address));
+    subscriptions[subscription] += _amount;
+  }
+
+  function validatePaymasterUserOp(
```

Comment:
does this need any authorization control? maybe only entrypoint should be able to call this?

Reply:
I wonder if token transfers and state mutations can be done outside of `validate*` function but doesn't seem like `postOp` contains the necessary context in the function signature

Reply:
Good catch. For the current implementation it definitely needs to be only callable by the entry point, otherwise someone could validate the hash and then not execute it, blocking the user operation from retries.

Reply:
And ditto on postOp